pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,317 @@
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+		return err
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) listAndSyncDataForGVK(ctx context.Context, gvk schema.GroupVersionKind, reader client.Reader) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) updateDatastore(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.replayTicker.C:
+			c.mu.RLock()
+			currentGVKsInAgg := watch.NewSet()
+			currentGVKsInAgg.Add(c.gvkAggregator.ListAllGVKs()...)
+			excluderChanged := c.excluderChanged
+			c.mu.RUnlock()
+
+			c.makeUpdates(ctx, currentGVKsInAgg, excluderChanged)
+		}
+	}
+}
+
+// listAndSyncData returns a set of gvks that were successfully listed and synced.
+func (c *CacheManager) listAndSyncData(ctx context.Context, gvks []schema.GroupVersionKind, reader client.Reader) *watch.Set {
+	gvksSuccessfullySynced := watch.NewSet()
+	for _, gvk := range gvks {
+		err := c.listAndSyncDataForGVK(ctx, gvk, c.reader)
+		if err != nil {
+			log.Error(err, "internal: error syncing gvks cache data")
+			// we don't remove this gvk as we will try to re-add it later
+			// we also don't return on this error to be able to list and sync
+			// other gvks in order to protect against a bad gvk.
+		} else {
+			gvksSuccessfullySynced.Add(gvk)
+		}
+	}
+	return gvksSuccessfullySynced
+}
+
+// makeUpdates performs a conditional wipe followed by a replay if necessary.
+func (c *CacheManager) makeUpdates(ctx context.Context, currentGVKsInAgg *watch.Set, excluderChanged bool) {
+	if c.watchedSet.Equals(currentGVKsInAgg) && !excluderChanged {
```

Comment:
> It could be easier to just write accessor functions that wrap the data store elements of the CacheManager with the obtaining of the mutex.

i think it's important to check point save the elements/ spec before we hand them off to `makeUpdates` in order to avoid raciness from updates to `AddSource` or `RemoveSource`. So we are trading off here some extensibility for correctness.

`unsetExcluderChanges` is needed to mark that we have processed the excluder change as soon as possible. Although maybe I can look to pull that out of `makeUpdates` somehow.

Reply:
n/a after move to list in the background